Dev container support

manandre opened the pull request:
### What is a dev container?
Introduce configuration files for setting up a dev environment for Npgsql inside a container, which can be hosted in any available service:
- [Docker](https://code.visualstudio.com/docs/remote/containers)
- [Docker for WSL2](https://code.visualstudio.com/docs/remote/wsl#_advanced-opening-a-wsl2-folder-in-a-container)
- [Codespaces](https://code.visualstudio.com/docs/remote/codespaces)
- (soon) [Codespaces for Github](https://github.com/features/codespaces/)
### Why a dev container?
Then, you can quickly setup a dev environment, without taking care about dependencies and their version, in order to develop a feature or a fix, or even to review a PR, in Visual Studio (Code) or even, for Codespaces, in your preferred browser! 😎
### How to configure a dev container?
We have a `.devcontainer` root folder containing:
- a `devcontainer.json` file for the container configuration
- a `Dockerfile` file detailing how to build the final image of the dev environment on top of the .Net 5.0 SDK and including the latest PostgreSQL server (as we do for the GitHub CI build)
### What is supported?
- .Net 5.0 target framework only
- Build and test (except plugins)
- Tests debugging
See screenshots from a "Codespaces from VS Code" setup:
![Npgsql-Codespaces](https://user-images.githubusercontent.com/2341261/82579040-84dd2180-9b8d-11ea-938d-b9b8ecd05351.png)
![Npgsql-Codespaces-debugger](https://user-images.githubusercontent.com/2341261/82579041-860e4e80-9b8d-11ea-8ea5-8d37d8f49ed4.png)
![Npgsql-Codespaces-debugger-exception](https://user-images.githubusercontent.com/2341261/82579043-86a6e500-9b8d-11ea-95f2-fc0cf7b0b56b.png)

roji commented:
Looks great to me! I personally just work on my Linux and manage my dependencies without using Docker, but anything that can make things easy for other people is a very good step forward.

* Any reason to install PG via apt, rather than just using [the official images](https://hub.docker.com/_/postgres?tab=tags)?
* I wonder if this could this also be used for the PG setup in our Github Actions CI config... As of now we have similar PostgreSQL setup happening in [build.yml](https://github.com/npgsql/npgsql/blob/dev/.github/workflows/build.yml), plus it's different for Linux and Windows. I'm not sure there's any value in running native PG on Windows (as opposed to the tests themselves which do need to run on Windows), this maybe this could simplify things there. A possible problem is that I don't see images for beta PG versions (13 beta1 was just released).

manandre commented:
I have changed the approach, I now use docker-compose to setup the dev environment from official .net sdk and postgresql images.
I use the `docker-entrypoint-initdb.d` folder to deploy our specific postgresql configuration for tests.
Moreover, I have added a pgadmin server to help for debugging but I do not manage to access it from the (codespaces) host 😞 If someone can help around port forwarding, (s)he is welcome 🤝 

This could potentially be reused in GitHub CI build workflow but I do not know the GitHub Actions API for now...

YohDeadfall commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,28 @@
+		"terminal.integrated.shell.linux": "/bin/bash",
+		"remote.extensionKind": {
+			"ms-azuretools.vscode-docker": "workspace"
+		  },
```
You have mixed spaces and tabs.

manandre replied:
Fixed!

YohDeadfall commented on .devcontainer/docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+    ports:
+      - 5050:80
+    links:
+      - db
```
Out of curiosity, it this required? I mean, what if I do not use pgAdmin at all, but have another tool to connect to the database?

roji replied:
I personally don't mind that it's there by default - I'm assuming the overhead is negligible for the environment...

manandre replied:
Yes, it is optional but I still do not manage to connect to it from dev host 😞
I have requested for some help from some devcontainer experts...

YohDeadfall commented on .vscode/settings.json:
```diff
@@ -0,0 +1,4 @@
+{
+    "omnisharp.defaultLaunchSolution": "Npgsql.sln",
+    "dotnet-test-explorer.testProjectPath": "**/*.Tests.csproj"
+}
```
Oh, you did it! Thank you!

One thing I would like to have is a styling setting in `.editorconfig` for JSON files, so all of them will use two spaces for indentation.

manandre replied:
Done.

manandre commented on .devcontainer/docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+    volumes:
+      - ..:/workspace:cached
+    # Overrides default command so things don't shut down after the process ends.
+    command: sleep infinity
```
Good point. I have tried it and it works.

manandre commented on .devcontainer/devcontainer.json:
```diff
@@ -0,0 +1,28 @@
+
+	"remoteEnv": {
+		"DeveloperBuild": "True",
+		"NPGSQL_TEST_DB": "Server=db;Username=npgsql_tests;Password=npgsql_tests;Database=npgsql_tests;Timeout=0;Command Timeout=0"
```
Removed.

roji commented on .vscode/extensions.json:
```diff
@@ -0,0 +1,11 @@
+{
+	// List of extensions which should be recommended for users of this workspace.
```
Are comments actually supported like this? They aren't in normal json...

manandre replied:
This is specific  to VS Code according to [docs](https://vscode-docs.readthedocs.io/en/stable/languages/json/)

roji commented on .vscode/extensions.json:
```diff
@@ -0,0 +1,11 @@
+{
```
Seems like this has tabs rather than the 2-space setting in .editorconfig

manandre replied:
Fixed.

roji commented:
Thanks @manandre!

Brar commented:
@manandre this is really a great idea and great work.
I actually didn't know the concept of dev containers yet.
I tried it out and also tried if I can solve the pgadmin connection issue.
I managed to get it running and also switched the db container to alpine since it's smaller/faster and we probably don't really care about the actual distro the db container runs.
Will file a PR in a second.
